packages/account-tree-controller/src/backup-and-sync/syncing/group.ts

```diff
@@ -0,0 +1,267 @@
+import { compareAndSyncMetadata } from './metadata';
+import type { AccountGroupMultichainAccountObject } from '../../group';
+import { backupAndSyncLogger } from '../../logger';
+import type { AccountWalletEntropyObject } from '../../wallet';
+import type { BackupAndSyncAnalyticsAction } from '../analytics';
+import { BackupAndSyncAnalyticsEvent } from '../analytics';
+import type { ProfileId } from '../authentication';
+import {
+  UserStorageSyncedWalletGroupSchema,
+  type BackupAndSyncContext,
+  type UserStorageSyncedWalletGroup,
+} from '../types';
+import {
+  pushGroupToUserStorage,
+  pushGroupToUserStorageBatch,
+} from '../user-storage/network-operations';
+import { getLocalGroupsForEntropyWallet } from '../utils';
+
+/**
+ * Creates a multichain account group.
+ *
+ * @param context - The sync context containing controller and messenger.
+ * @param entropySourceId - The entropy source ID.
+ * @param groupIndex - The group index.
+ * @param profileId - The profile ID for analytics.
+ * @param analyticsAction - The analytics action to log.
+ */
+export const createMultichainAccountGroup = async (
+  context: BackupAndSyncContext,
+  entropySourceId: string,
+  groupIndex: number,
+  profileId: ProfileId,
+  analyticsAction: BackupAndSyncAnalyticsAction,
+) => {
+  try {
+    // This will be idempotent so we can create the group even if it already exists
+    await context.messenger.call(
+      'MultichainAccountService:createMultichainAccountGroup',
+      {
+        entropySource: entropySourceId,
+        groupIndex,
+      },
+    );
+
+    context.emitAnalyticsEventFn({
+      action: analyticsAction,
+      profileId,
+    });
+  } catch (error) {
+    backupAndSyncLogger(
+      `Failed to create group ${groupIndex} for entropy ${entropySourceId}:`,
+      // istanbul ignore next
+      error instanceof Error ? error.message : String(error),
+    );
+    throw error;
+  }
+};
+
+/**
+ * Creates local groups from user storage groups.
+ *
+ * @param context - The sync context containing controller and messenger.
+ * @param groupsFromUserStorage - Array of groups from user storage.
+ * @param entropySourceId - The entropy source ID.
+ * @param profileId - The profile ID for analytics.
+ */
+export async function createLocalGroupsFromUserStorage(
+  context: BackupAndSyncContext,
+  groupsFromUserStorage: UserStorageSyncedWalletGroup[],
+  entropySourceId: string,
+  profileId: ProfileId,
+): Promise<void> {
+  const numberOfAccountGroupsToCreate = Math.max(
+    ...groupsFromUserStorage.map((g) => g.groupIndex),
+  );
+
+  for (
+    let groupIndex = 0;
+    groupIndex <= numberOfAccountGroupsToCreate;
+    groupIndex++
+  ) {
+    try {
+      await createMultichainAccountGroup(
```

Comment:
Nit: We could maybe add a small comment to mention idempotency? Something like:

```suggestion
      // Creating multichain account group is idempotent, so we can safely
      // re-create every groups starting from 0.
      await createMultichainAccountGroup(
```